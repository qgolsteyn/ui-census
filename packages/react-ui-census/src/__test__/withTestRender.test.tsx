import React from "react";
import { createTestRender } from "..";

const b = createTestRender(
  {
    div: (target, query: { id?: string } = {}) => {
      const elements = Array.from(target.querySelectorAll("div"));

      return elements.filter((element) => {
        if (query.id !== undefined) {
          return element.id === query.id;
        } else {
          return true;
        }
      });
    },
  },
  {
    div: {
      text: (element) => element.textContent,
    },
  }
);

test("it handles a basic query", () => {
  const doc = b(<div>Test 1</div>);

  expect(doc.div.text).toBe("Test 1");
});

test("it handles a basic query by id", () => {
  const doc = b(<div id="test">Test 2</div>);

  expect(doc.div.q("test").text).toBe("Test 2");
});

test("it handles a basic query by prop", () => {
  const doc = b(<div id="test">Test 3</div>);

  expect(doc.div.q({ id: "test" })[0].text).toBe("Test 3");
});

test("it handles a query by prop with mapping", () => {
  const doc = b(
    <>
      <div>1</div>
      <div>2</div>
      <div>3</div>
    </>
  );

  expect(doc.div.q({}).map((node) => node.text)).toStrictEqual(["1", "2", "3"]);
});

test("it supports snapshot testing", () => {
  const doc = b(<div id="test">Test 4</div>);

  expect(doc.div).toMatchSnapshot();
  expect(doc.div.q("test")).toMatchSnapshot();
  expect(doc.div.q({ id: "test" })).toMatchSnapshot();
});

test("it throws when accessing default query where multiple elements are present", () => {
  const doc = b(
    <>
      <div>1</div>
      <div>2</div>
      <div>3</div>
    </>
  );

  expect(() => doc.div.text).toThrow();
});

test("it throws when accessing id query where multiple elements are present", () => {
  const doc = b(
    <>
      <div>1</div>
      <div>2</div>
      <div>3</div>
    </>
  );

  expect(() => doc.div.q("test").text).toThrow();
});
