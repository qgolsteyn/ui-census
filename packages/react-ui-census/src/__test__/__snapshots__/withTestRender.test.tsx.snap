// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it supports snapshot testing 1`] = `
Object {
  "text": "Test 4",
}
`;

exports[`it supports snapshot testing 2`] = `
Object {
  "text": "Test 4",
}
`;

exports[`it supports snapshot testing 3`] = `
Array [
  Object {
    "text": "Test 4",
  },
]
`;
